internalshared/configutil/listener.go

```diff
@@ -279,27 +292,28 @@ func (l *Listener) parseChrootNamespaceSettings() error {
 // The state of the listener will be modified.
 func (l *Listener) parseType(fallback string) error {
 	switch {
-	case l.Type != "":
+	case l.Type.String() != "":
 	case fallback != "":
 	default:
 		return errors.New("listener type must be specified")
 	}
 
 	// Use type if available, otherwise fall back.
-	result := l.Type
-	if result == "" {
-		result = fallback
+	rawType := l.Type
+	if rawType == "" {
+		rawType = listenerType(fallback)
 	}
-	result = strings.ToLower(result)
+
+	parsedType := listenerType(strings.ToLower(rawType.String()))
 
 	// Sanity check the values
-	switch result {
-	case "tcp", "unix":
+	switch parsedType {
+	case TCP, Unix:
 	default:
-		return fmt.Errorf("unsupported listener type %q", result)
+		return fmt.Errorf("unsupported listener type %q", parsedType.String())
```

Comment:
I don't think the `.String()` is necessary. If `parsedType` is simply provided as the argument, the `%q` handles that value correctly and writes in the string value in quotation marks as desired.